File: frontend/src/engine/comminution.ts
/**
 * Breakage matrices, the Whiten crusher and the energy-specific ball-mill operator (port of
 * engine/comminution.py; theory in docs/methodologies/02_crushing.md and 03_grinding-circuit.md).
 */
import { constant } from './constants';
import { grid, type Vec } from './grid';
import { eye, matmul, scaleColumns, solve, type Mat } from './linalg';
import type { Crusher, Mill } from './model';

/** Strictly lower-triangular b[i][j]: fraction of broken class-j mass reporting to class i. */
export function breakageMatrix(beta0: number, beta1: number, beta2: number): Mat {
  const g = grid();
  const n = g.n;
  const data = new Float64Array(n * n);
  for (let j = 0; j < n - 1; j += 1) {
    const cumulative: number[] = [];
    for (let i = j + 1; i < n; i += 1) {
      const ratio = g.upper[i] / g.upper[j + 1];
      cumulative.push(beta0 * Math.pow(ratio, beta1) + (1.0 - beta0) * Math.pow(ratio, beta2));
    }
    for (let k = 0; k < cumulative.length; k += 1) {
      const value = k < cumulative.length - 1 ? cumulative[k] - cumulative[k + 1] : cumulative[k];
      data[(j + 1 + k) * n + j] = value;
    }
  }
  return { n, data };
}

/** Probability that a particle of each class enters the breakage zone. */
export function whitenClassification(cssUm: number, crusher: Crusher): Vec {
  const g = grid();
  const k1 = crusher.k1_css * cssUm;
  const k2 = crusher.k2_css * cssUm;
  const c = new Float64Array(g.n);
  for (let i = 0; i < g.n; i += 1) {
    const size = g.size[i];
    if (size < k1) c[i] = 0.0;
    else if (size > k2) c[i] = 1.0;
    else c[i] = 1.0 - Math.pow(Math.min(1.0, Math.max(0.0, (k2 - size) / (k2 - k1))), crusher.k3);
  }
  c[g.n - 1] = 0.0;
  return c;
}

/** Whiten crusher product p = (I - C)(I - B C)^-1 f for one mineral's feed vector. */
export function crush(feed: Vec, cssUm: number, crusher: Crusher): Vec {
  const b = breakageMatrix(crusher.beta0, crusher.beta1, crusher.beta2);
  const c = whitenClassification(cssUm, crusher);
  const n = grid().n;
  const bc = scaleColumns(b, c);
  const system = eye(n);
  for (let k = 0; k < n * n; k += 1) system.data[k] -= bc.data[k];
  const internal = solve(system, feed);
  const out = new Float64Array(n);
  for (let i = 0; i < n; i += 1) out[i] = (1.0 - c[i]) * internal[i];
  return out;
}

/** Energy-specific selection function S^E (t/kWh) by class; the pan does not break. */
export function selectionEnergy(mill: Mill, workIndex: number): Vec {
  const g = grid();
  const s = new Float64Array(g.n);
  for (let i = 0; i < g.n; i += 1) {
    const raw = mill.alpha0 * (mill.reference_work_index_kwh_t / workIndex) * Math.pow(g.size[i], mill.alpha1);
    s[i] = raw / (1.0 + Math.pow(g.size[i] / mill.critical_size_um, mill.alpha2));
  }
  s[g.n - 1] = 0.0;
  return s;
}

/** T^-1(e) = prod_k (I + e f_k D) for one mineral, expanded once in powers of D. */
export class MillOperator {
  readonly d: Mat;
  readonly d2: Mat;
  readonly d3: Mat;
  readonly c1: number;
  readonly c2: number;
  readonly c3: number;

  constructor(selection: Vec, b: Mat) {
    const n = grid().n;
    const iMinusB = eye(n);
    for (let k = 0; k < n * n; k += 1) iMinusB.data[k] -= b.data[k];
    this.d = scaleColumns(iMinusB, selection);
    this.d2 = matmul(this.d, this.d);
    this.d3 = matmul(this.d2, this.d);
    const [f1, f2, f3] = constant<number[]>('mill.mixer_fractions');
    this.c1 = f1 + f2 + f3;
    this.c2 = f1 * f2 + f1 * f3 + f2 * f3;
    this.c3 = f1 * f2 * f3;
  }

  inverse(energyPerPass: number): Mat {
    const e = energyPerPass;
    const n = this.d.n;
    const a1 = this.c1 * e;
    const a2 = this.c2 * e * e;
    const a3 = this.c3 * e * e * e;
    // the same order of additions as the Python engine: ((I + a1 D) + a2 D^2) + a3 D^3
    const data = eye(n).data;
    for (let k = 0; k < n * n; k += 1) data[k] = ((data[k] + a1 * this.d.data[k]) + a2 * this.d2.data[k]) + a3 * this.d3.data[k];
    return { n, data };
  }
}

/** Bond specific energy (kWh/t) for a reduction from F80 to P80 (GMG01-MP-2021). */
export function bondEnergy(workIndex: number, f80Um: number, p80Um: number): number {
  const k = constant('bond.coefficient');
  return workIndex * (k / Math.sqrt(p80Um) - k / Math.sqrt(f80Um));
}

export function operatingWorkIndex(energy: number, f80Um: number, p80Um: number): number {
  const k = constant('bond.coefficient');
  return energy / (k / Math.sqrt(p80Um) - k / Math.sqrt(f80Um));
}

File: data-pipeline/pipeline/engine/comminution.py
"""Breakage matrices, the Whiten crusher and the energy-specific ball-mill operator.

Sources (docs/frameworks/comminution): Whiten crusher form as reproduced in Andrejev et al. (2021),
doi:10.3390/min11111256; energy-specific selection after Herbst and Fuerstenau (1980),
doi:10.1016/0301-7516(80)90034-4, in the Moly-Cop form; Austin breakage function; three perfect
mixers after Austin, Klimpel and Luckie (1984).
"""
from __future__ import annotations

import numpy as np

from .constants import constant
from .grid import grid
from .model import Crusher, Mill


def breakage_matrix(beta0: float, beta1: float, beta2: float) -> np.ndarray:
    """Strictly lower-triangular b[i, j]: fraction of broken class-j mass reporting to class i.

    ``B(i, j) = beta0 (x_i/x_{j+1})^beta1 + (1 - beta0)(x_i/x_{j+1})^beta2`` is the cumulative
    fraction of primary products of class j finer than the upper bound ``x_i`` of class i; the
    column of every breakable class sums to one and the pan column is zero.
    """
    g = grid()
    n = g.n
    b = np.zeros((n, n))
    for j in range(n - 1):
        ratio = g.upper[j + 1:] / g.upper[j + 1]
        cumulative = beta0 * np.power(ratio, beta1) + (1.0 - beta0) * np.power(ratio, beta2)
        column = np.empty(n - j - 1)
        column[:-1] = cumulative[:-1] - cumulative[1:]
        column[-1] = cumulative[-1]
        b[j + 1:, j] = column
    return b


def whiten_classification(css_um: float, crusher: Crusher) -> np.ndarray:
    """Probability that a particle of each class enters the breakage zone."""
    g = grid()
    k1 = crusher.k1_css * css_um
    k2 = crusher.k2_css * css_um
    size = g.size
    c = np.where(size < k1, 0.0, np.where(size > k2, 1.0, 1.0 - np.power(np.clip((k2 - size) / (k2 - k1), 0.0, 1.0), crusher.k3)))
    c[-1] = 0.0
    return c


def crush(feed: np.ndarray, css_um: float, crusher: Crusher) -> np.ndarray:
    """Whiten crusher product ``p = (I - C)(I - B C)^-1 f`` for one mineral's feed vector."""
    b = breakage_matrix(crusher.beta0, crusher.beta1, crusher.beta2)
    c = whiten_classification(css_um, crusher)
    n = grid().n
    internal = np.linalg.solve(np.eye(n) - b * c[None, :], feed)
    return (1.0 - c) * internal


def selection_energy(mill: Mill, work_index: float) -> np.ndarray:
    """Energy-specific selection function S^E (t/kWh) by class; the pan does not break."""
    g = grid()
    s = mill.alpha0 * (mill.reference_work_index_kwh_t / work_index) * np.power(g.size, mill.alpha1)
    s = s / (1.0 + np.power(g.size / mill.critical_size_um, mill.alpha2))
    s[-1] = 0.0
    return s


class MillOperator:
    """``T^-1(e) = prod_k (I + e f_k D)`` for one mineral, expanded once in powers of D."""

    def __init__(self, selection: np.ndarray, b: np.ndarray) -> None:
        n = grid().n
        self.d = (np.eye(n) - b) * selection[None, :]
        self.d2 = self.d @ self.d
        self.d3 = self.d2 @ self.d
        f1, f2, f3 = (float(v) for v in constant("mill.mixer_fractions"))
        self.c1 = f1 + f2 + f3
        self.c2 = f1 * f2 + f1 * f3 + f2 * f3
        self.c3 = f1 * f2 * f3
        self.eye = np.eye(n)

    def inverse(self, energy_per_pass: float) -> np.ndarray:
        e = energy_per_pass
        return self.eye + (self.c1 * e) * self.d + (self.c2 * e * e) * self.d2 + (self.c3 * e * e * e) * self.d3


def bond_energy(work_index: float, f80_um: float, p80_um: float) -> float:
    """Bond specific energy (kWh/t) for a reduction from F80 to P80 (GMG01-MP-2021)."""
    k = float(constant("bond.coefficient"))
    return work_index * (k / np.sqrt(p80_um) - k / np.sqrt(f80_um))


def operating_work_index(energy: float, f80_um: float, p80_um: float) -> float:
    k = float(constant("bond.coefficient"))
    return energy / (k / np.sqrt(p80_um) - k / np.sqrt(f80_um))
